harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+            self._init_pynvml()
+        else:
+            logging.warning("pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        """
+        Call the NVML library methods we'll be using to validate that they are accessible.
+        Sometimes NVML initializes successfully but individual device methods will fail.
+        """
+        try:
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                uuid = pynvml.nvmlDeviceGetUUID(handle)
+                pynvml.nvmlDeviceGetMemoryInfo(handle)
+                pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml_device_handles[uuid] = handle
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
+            logging.error(f"Error initializing pynvml: {e}")
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        metrics = {}
+
+        for uuid, handle in self._pynvml_device_handles.items():
+            gpu_util = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
+            free_memory = pynvml.nvmlDeviceGetMemoryInfo(handle).free
+            metrics.update(
+                {
+                    uuid: {
+                        "gpu_util": gpu_util,
+                        "gpu_free_memory": free_memory,
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: typing.Dict[str, typing.Any],
+        timestamp: datetime.datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _Collector(threading.Thread):
+    def __init__(
+        self,
+        metrics_queue: queue.Queue,
+        collection_interval: int = 1,
+        max_collection_s: int = 60 * 30,
+    ):
+        self._collection_interval = collection_interval
+        self._metrics_queue = metrics_queue
+        self._max_collection_s = max_collection_s
+        self._metric_collectors = [
+            _GPU(),
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(None)
+                return
+
+            next_collection_ts = time.time() + self._collection_interval
+            self._collect_metrics()
+            wait_ts = max(next_collection_ts - time.time(), 0)
```

Comment:
That the cadence can be variable means that the metrics should never be counts. What's the best way to convey this information?

Reply:
I see in `bindings` it's called `avgMetrics`. That's an idea.

Reply:
i feel like this should be pretty obvious if someone is implementing a new metric, and such a metric seems rare. i'm having a hard time thinking of a metric that is "number of things that happened since the last time we measured" that wouldn't be some sort of _per_second calculation.